src/main/java/edu/texas/social_computing/hospitals/RHRP.java

```diff
@@ -0,0 +1,134 @@
+     * - Assumes that both partners have same pref list size
+     */
+
+    public static Matching run(HospitalTable hospitalTable,
+                               ResidentTable residentTable,
+                               Queue<Resident> freeResidents,
+                               Matching existingMatching) {
+        Matching m = existingMatching;
+
+        while (!freeResidents.isEmpty()) {
+            Resident currentResident = freeResidents.poll();
+            if (currentResident.getPreferences().isEmpty()) {
+                if (!currentResident.hasPartner()) {
+                    continue;
+                }
+                if (residentTable.getResidentById(currentResident.getPartnerId()).getPreferences().isEmpty()) {
+                    continue;
+                }
+                // resident and partner have to have the same pref list size
+                assert (false);
+            }
+
+            Hospital hospital = hospitalTable.getHospitalById(currentResident.getPreferences()
+                    .get(residentTable.getResidentRankProgress(currentResident)));
+            residentTable.incrementResidentRankProgress(currentResident);
+
+            tryToMatch(hospital,currentResident,residentTable,freeResidents,m);
+
+            // if the currentResident managed to get assigned then try to assign partner if he has one
+            if (m.hasAssignment(currentResident) && currentResident.hasPartner()) {
+                tryToMatch(hospital,residentTable.getResidentById(currentResident.getPartnerId()),
+                        residentTable,freeResidents,m);
+            }
+
+            // if currentResident has more applying to do, put back in Q
+            if (!m.hasAssignment(currentResident)) {
+                freeResidents.add(currentResident);
+                if (currentResident.hasPartner()) {
+                    freeResidents.add(residentTable.getResidentById(currentResident.getPartnerId()));
+                }
+            }
+        }
+        return m;
+    }
+
+    public static Matching run(HospitalTable hospitalTable,
+                               ResidentTable residentTable,
+                               Queue<Resident> freeResidents) {
+        return run(hospitalTable, residentTable, freeResidents, new Matching());
+    }
+
+    public static void main(String[] args) throws FileNotFoundException {
+        List<Resident> residents = FileImporter.importResidents(args[0]);
+        List<Hospital> hospitals = FileImporter.importHospitals(args[1]);
+        System.out.println("loaded Files");
+        System.out.println("residents: " + Integer.toString(residents.size()));
+        System.out.println("hospitals: " + Integer.toString(hospitals.size()));
+
+        // make lookup tables
+        HospitalTable hospitalTable = HospitalTable.create(hospitals);
+        ResidentTable residentTable = ResidentTable.create(residents);
+        System.out.println("generated tables");
+
+        // call HRP
```

Comment:
right before you run, since i know you really want to assert, you can:
`assert(hasValidCouples(residentTable);`

Reply:
Prefer throwing exceptions to `asserts`